# PSO Implementation

---


### 1. Import Libraries

---

In [10]:
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
from pso_class import ParticleSwarmOptimization  


---

### 2. Define the Objective Function

In [11]:
def f(x, y):
    """
    Objective function with multiple local minima.
    
    This function combines quadratic terms and sinusoidal components to create
    a complex landscape with multiple local minima and one global minimum.
    
    Parameters:
    -----------
    x, y : array-like
        Input coordinates
        
    Returns:
    --------
    array-like
        Function values at the given coordinates
    """
    return (x - 3.14) ** 2 + (y - 2.72) ** 2 + np.sin(3 * x + 1.41) + np.sin(4 * y - 1.73)


---

### 3. Compute and visualize the function landscape 

In [12]:
print("1. Global Minimum Calculation")
# Create a grid of points covering the search space [0,5]x[0,5]
x_vals = np.linspace(0, 5, 1000)
y_vals = np.linspace(0, 5, 1000)
x_grid, y_grid = np.meshgrid(x_vals, y_vals)
z_grid = f(x_grid, y_grid)

# Find the global minimum
min_idx = np.argmin(z_grid)
y_min, x_min = np.unravel_index(min_idx, z_grid.shape)
x_min, y_min = x_vals[x_min], y_vals[y_min]
min_value = f(x_min, y_min)

print(f"   Global minimum found at x={x_min:.6f}, y={y_min:.6f}")
print(f"   Minimum value: f(x,y)={min_value:.6f}")

1. Global Minimum Calculation
   Global minimum found at x=3.183183, y=3.128128
   Minimum value: f(x,y)=-1.808306


---

### 4. Initizalide the PSO Algorithm

In [13]:
print("\n2. PSO Initialization and Execution")
# Define PSO parameters
n_particles = 20  # Number of particles in the swarm
w = 0.8           # Inertia weight
c1 = c2 = 0.1     # Cognitive and social parameters
bounds = (0, 5)   # Search space bounds
seed = 100        # Random seed for reproducibility
iterations = 50   # Number of iterations to run

print(f"   Number of particles: {n_particles}")
print(f"   Parameters: w={w}, c1={c1}, c2={c2}")
print(f"   Search space bounds: {bounds}")
print(f"   Number of iterations: {iterations}")

# Create PSO instance
pso = ParticleSwarmOptimization(
    objective_function=f,
    n_particles=n_particles,
    dimensions=2,
    bounds=bounds,
    w=w,
    c1=c1,
    c2=c2,
    seed=seed
)

# Run the optimization
print("\n3. Running Optimization")
gbest, gbest_obj = pso.optimize(iterations)


2. PSO Initialization and Execution
   Number of particles: 20
   Parameters: w=0.8, c1=0.1, c2=0.1
   Search space bounds: (0, 5)
   Number of iterations: 50

3. Running Optimization


---

### 5. Display results

In [14]:
print("\n4. Results:")
print(f"   PSO found best solution at:")
print(f"      x = {gbest[0]:.6f}, y = {gbest[1]:.6f}")
print(f"      f(x,y) = {gbest_obj:.6f}")
print(f"\n   True global minimum:")
print(f"      x = {x_min:.6f}, y = {y_min:.6f}")
print(f"      f(x,y) = {min_value:.6f}")

# Calculate the error
distance_error = np.sqrt((gbest[0] - x_min)**2 + (gbest[1] - y_min)**2)
value_error = abs(gbest_obj - min_value)
print(f"\n   Error Analysis:")
print(f"      Distance error: {distance_error:.6f}")
print(f"      Function value error: {value_error:.6f}")


4. Results:
   PSO found best solution at:
      x = 3.185418, y = 3.129725
      f(x,y) = -1.808352

   True global minimum:
      x = 3.183183, y = 3.128128
      f(x,y) = -1.808306

   Error Analysis:
      Distance error: 0.002746
      Function value error: 0.000046


---

### 6. Convergence Analysis

In [15]:
print("\n5. Convergence Analysis")
history = pso.get_history()
gbest_values = history['gbest_obj']

# Print values at key iterations
print("   Objective value by iteration:")
iterations_to_show = [0, 4, 9, 19, 49] if len(gbest_values) >= 50 else list(range(0, len(gbest_values), max(1, len(gbest_values)//5)))
for i in iterations_to_show:
    if i < len(gbest_values):
        print(f"      Iteration {i}: {gbest_values[i]:.6f}")

# Calculate improvement
if len(gbest_values) > 1:
    total_improvement = gbest_values[0] - gbest_values[-1]
    improvement_percent = (total_improvement / gbest_values[0]) * 100
    print(f"\n   Total improvement: {total_improvement:.6f} ({improvement_percent:.2f}%)")
    
    # Check if converged
    if len(gbest_values) >= 10:
        recent_improvement = gbest_values[-10] - gbest_values[-1]
        if recent_improvement < 1e-6:
            print("   Convergence status: Solution has converged (minimal improvement in last 10 iterations)")
        else:
            print("   Convergence status: Solution still improving")
    else:
        print("   Convergence status: Not enough iterations to determine")


5. Convergence Analysis
   Objective value by iteration:
      Iteration 0: 0.940058
      Iteration 4: -0.219024
      Iteration 9: -1.764420
      Iteration 19: -1.808225
      Iteration 49: -1.808352

   Total improvement: 2.748410 (292.37%)
   Convergence status: Solution has converged (minimal improvement in last 10 iterations)


---

# PSO Algorithm Results Analysis

## Comparison of Results

| Metric | PSO Solution | True Global Minimum | Difference |
|--------|-------------|---------------------|------------|
| x-coordinate | 3.185418 | 3.183183 | 0.002235 |
| y-coordinate | 3.129725 | 3.128128 | 0.001597 |
| Function value | -1.808352 | -1.808306 | 0.000046 |
| Euclidean distance | - | - | 0.002746 |

## Convergence Analysis

| Iteration | Objective Value | Improvement from Previous |
|-----------|----------------|--------------------------|
| 0 | 0.940058 | - |
| 4 | -0.219024 | 1.159082 |
| 9 | -1.764420 | 1.545396 |
| 19 | -1.808225 | 0.043805 |
| 49 | -1.808352 | 0.000127 |

Total improvement: 2.748410 (292.37%)

## Conclusions

1. **Accuracy**: The PSO algorithm successfully found the global minimum of the non-linear function with high precision. The difference between the PSO solution and the true global minimum is negligible (function value error of only 0.000046).

2. **Convergence**: The algorithm demonstrated excellent convergence properties:
   - Rapid improvement in early iterations (particularly between iterations 0-9)
   - Fine-tuning in middle iterations (9-19)
   - Minimal improvement in later iterations, indicating proper convergence

3. **Efficiency**: The algorithm converged to a near-optimal solution within just 20 iterations, with only minimal improvements thereafter.

4. **Stability**: The solution stabilized in the later iterations, confirming that the algorithm successfully settled at the global minimum rather than oscillating between local minima.

5. **Parameter Effectiveness**: The chosen parameters (w=0.8, c1=c2=0.1, n_particles=20) proved very effective for this particular optimization problem, balancing exploration and exploitation.